### PR TITLE
fixed Nacos Config correct order #987 on master branch

### LINKED ISSUE #987: Nacos Config with yaml file not in sequence
我们鼓励使用英文，如果不能直接使用，可以使用翻译软件，您仍旧可以保留中文原文。
We recommend using English. If you are non-native English speaker, you can use the translation software. 

**Which Component**
Nacos Config

**Describe the bug**
spring-cloud-starter-alibaba-nacos-config with yaml file convert data will not in sequence，
my project (a basic spring cloud project ) change to use spring-cloud-with-alibaba,use Nacos to alternative Eureka and Config, I found yml file can't convert in sequence.
so make my shiro config will wrong

is the same issue on Nacos Project : https://github.com/alibaba/nacos/issues/1910

**To Reproduce**

**Expected behavior**

**Screenshots**

**Additional context**
MacOS
Spring Cloud 2.1.0
jdk1.8

### PR DESCRIPTION
**Pull request by forfuns**
### Describe what this PR does / why we need it

fix bug

### Does this pull request fix one issue?

Fixes #987 

### Describe how to verify it

use nacos config client and see that configuration sequence

### Special notes for reviews

for master branch( next version is `2.2.0` ?),but for `2.1.0.RELEASE` is on `[develop-2.1.x-fix](https://github.com/forfuns/spring-cloud-alibaba/tree/develop-2.1.x-fix)`,because the code has been refactor

### CODE REVIEW

**Comment by zkzlx** [comment]
@forfuns  First of all, thanks for your suggestion , but i think  there's a better way.
 Simply using an ordered collection to receive the parsed results, .
> This eliminates the overhead of switching back and forth between `Map` and `Properties`  and need to customize  `Properties`  and simplifies the code .

**Comment by zkzlx** [comment]
Here's another tip :
If you define a class with a special function, it is better to narrow its scope of use so that it can be effectively protected

**Comment by forfuns** [comment]
@zkzlx `YamlMapFactoryBean` parsed results is not a ordered collection,but `YamlMapFactoryBean` is use `LinkedHashMap`, that can be ordered, then convert to ordered properties.on otherwise, need rewrite a new yaml parser

**Comment by zkzlx** [comment]
> @zkzlx `YamlMapFactoryBean` parsed results is not a ordered collection,but `YamlMapFactoryBean` is use `LinkedHashMap`, that can be ordered, then convert to ordered properties.on otherwise, need rewrite a new yaml parser

You can refer to `YamlMapFactoryBean` , it is already paresed for you , just need to make a converter.
refer to `org.springframework.beans.factory.config.YamlProcessor#getFlattenedMap`

**Comment by forfuns** [comment]
fixed on new commit:cde1fe5

File: spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java [hunk 1]
```diff
@@ -36,13 +44,19 @@ protected Properties doParse(String data) throws IOException {
...
-			if (arr.length != 2) {
-				throw new IOException("error properties format");
+			sourceArr = line.split("=");
+			if (sourceArr.length < 2) {
+				log.debug("ignore no properties format line :{}", line);
 			}
-
-			properties.put(arr[0], arr[1]);
+			key = sourceArr[0].trim();
+			valueArr = Arrays.copyOfRange(sourceArr, 1, sourceArr.length);
```
Comment (zkzlx) [inline spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java]: this a piece of code use `substring()` and `indexOf()` is better done with it

File: spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java [hunk 2]
```diff
@@ -16,23 +16,48 @@
...
 
+	private static final Logger log = LoggerFactory
+			.getLogger(NacosDataPropertiesParser.class);
+
 	public NacosDataPropertiesParser() {
 		super("properties");
 	}
 
 	@Override
 	protected Properties doParse(String data) throws IOException {
```
Comment (zkzlx) [inline spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java]: It is recommended to replace `Proterties` with `LinkedHashMap` , unnecessary conversion of data types is reduced , so you need to modify the affected code of data flow .

File: spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataYamlParser.java [hunk 3]
```diff
@@ -32,9 +40,63 @@ public NacosDataYamlParser() {
 
 	@Override
 	protected Properties doParse(String data) {
```
Comment (zkzlx) [inline spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataYamlParser.java]: It is recommended to replace Proterties with LinkedHashMap , unnecessary conversion of data types is reduced , so you need to modify the affected code of data flow . Other parsers need to do the same .

**Comment by forfuns** [comment]
兄弟，你review的那两段代码可以用中文发邮件告诉我吗，英语水平有点烂，勉强能阅读，但容易理解错
| |
Bin
|
|
my-tangjianbin@163.com
|
签名由网易邮箱大师定制


On 10/14/2019 16:06，zkzlx<notifications@github.com> wrote：

@zkzlx commented on this pull request.

In spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java:

>  			}
-
-			properties.put(arr[0], arr[1]);
+			key = sourceArr[0].trim();
+			valueArr = Arrays.copyOfRange(sourceArr, 1, sourceArr.length);


this a piece of code use substring() and indexOf() is better done with it

File: spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java [hunk 4]
```diff
@@ -16,23 +16,48 @@
...
+		BufferedReader br = new BufferedReader(reader);
+
+		String key;
+		String value;
+		String[] sourceArr;
+		String[] valueArr;
+		for (String line = br.readLine(); line != null; line = br.readLine()) {
+			sourceArr = line.split("=");
+			if (sourceArr.length < 2) {
+				log.debug("ignore no properties format line :{}", line);
```
Comment (zkzlx) [inline spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java]: There should be a bug here , missing keyword of `conitnue`  .In addition, the logging level here should be `warn` that is better

File: spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java (hunk 1 above)
Comment (forfuns) [reply spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java]: > this a piece of code use `substring()` and `indexOf()` is better done with it

how  can I use substring and indexOf at array?

File: spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java (hunk 2 above)
Comment (forfuns) [reply spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java]: > It is recommended to replace `Proterties` with `LinkedHashMap` , unnecessary conversion of data types is reduced , so you need to modify the affected code of data flow .

this PR only for fix orderly properties not for refactor

File: spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataYamlParser.java (hunk 3 above)
Comment (forfuns) [reply spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataYamlParser.java]: > It is recommended to replace `Proterties ` with `LinkedHashMap `, unnecessary conversion of data types is reduced , so you need to modify the affected code of data flow . Other parsers need to do the same .

this PR only for fix orderly properties not for refactor

File: spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java (hunk 4 above)
Comment (forfuns) [reply spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java]: > There should be a bug here , missing keyword of `conitnue` .In addition, the logging level here should be `warn` that is better

I thought about this situation is part of properties format wrong , but we ignore it and make other pass , that include "annotation" , if you suggest level should be `warn`,so I rewrite this code distinguish between "annotation" and "ignore"

File: spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java (hunk 1 above)
Comment (zkzlx) [reply spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java]: > > this a piece of code use `substring()` and `indexOf()` is better done with it
> 
> how can I use substring and indexOf at array?

e.g line.substring(line.indexOf("=") + 1)

Comment (forfuns) [reply spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataPropertiesParser.java]: complete

File: spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataYamlParser.java (hunk 3 above)
Comment (fangjian0423) [reply spring-cloud-alibaba-nacos-config/src/main/java/com/alibaba/cloud/nacos/parser/NacosDataYamlParser.java]: i hope you can do some refactors, don't want to merge some bad codes. thanks

**Pull request closed without merging**